File: bookforge/tools/workspace_lock.py
"""Thread-safe workspace operations with locking for parallel execution.

This module provides a locking mechanism for manifest updates to prevent
race conditions when multiple videos are processed concurrently.
"""

from __future__ import annotations

import asyncio
from pathlib import Path
from typing import Any

from bookforge.schemas import ChannelManifest
from bookforge.tools.workspace import Workspace

# Global lock per workspace (keyed by workspace root path)
_workspace_locks: dict[str, asyncio.Lock] = {}


def get_workspace_lock(workspace_root: Path) -> asyncio.Lock:
    """Get or create a lock for the given workspace.
    
    This ensures that manifest updates for the same workspace are serialized,
    preventing race conditions when multiple videos are processed in parallel.
    """
    key = str(workspace_root.resolve())
    if key not in _workspace_locks:
        _workspace_locks[key] = asyncio.Lock()
    return _workspace_locks[key]


async def safe_update_video(
    ws: Workspace, video_id: str, **changes: Any
) -> ChannelManifest:
    """Thread-safe video status update with automatic locking.
    
    This function wraps Workspace.update_video() with an asyncio lock to
    prevent concurrent modification of the manifest file.
    
    Args:
        ws: The Workspace instance
        video_id: Video ID to update
        **changes: Status fields to update (status, error, etc.)
    
    Returns:
        Updated manifest
    
    Example:
        await safe_update_video(ws, "abc123", status="verified", error="")
    """
    lock = get_workspace_lock(ws.root)
    async with lock:
        # Perform the update atomically
        return ws.update_video(video_id, **changes)


async def safe_load_manifest(ws: Workspace) -> ChannelManifest:
    """Thread-safe manifest loading.
    
    While loading is typically safe without a lock, this function is provided
    for consistency and can be used when you need to ensure no writes happen
    during a read operation.
    """
    lock = get_workspace_lock(ws.root)
    async with lock:
        return ws.load_manifest()


async def safe_save_manifest(ws: Workspace, manifest: ChannelManifest) -> None:
    """Thread-safe manifest saving.
    
    Use this when you need to save an entire manifest that was modified
    outside of update_video().
    """
    lock = get_workspace_lock(ws.root)
    async with lock:
        ws.save_manifest(manifest)

File: bookforge-go/tools/workspace_lock.go
package tools

import (
	"sync"
)

// WorkspaceLock provides thread-safe access to workspace operations
// to prevent race conditions during parallel video processing
type WorkspaceLock struct {
	mu sync.Mutex
}

// NewWorkspaceLock creates a new workspace lock
func NewWorkspaceLock() *WorkspaceLock {
	return &WorkspaceLock{}
}

// Global lock manager for workspaces (keyed by workspace root)
var (
	workspaceLocks   = make(map[string]*WorkspaceLock)
	workspaceLocksRW sync.RWMutex
)

// GetWorkspaceLock returns or creates a lock for the given workspace path
func GetWorkspaceLock(workspaceRoot string) *WorkspaceLock {
	// Fast path: read lock to check if exists
	workspaceLocksRW.RLock()
	lock, exists := workspaceLocks[workspaceRoot]
	workspaceLocksRW.RUnlock()

	if exists {
		return lock
	}

	// Slow path: write lock to create
	workspaceLocksRW.Lock()
	defer workspaceLocksRW.Unlock()

	// Double-check in case another goroutine created it
	if lock, exists := workspaceLocks[workspaceRoot]; exists {
		return lock
	}

	// Create new lock
	lock = NewWorkspaceLock()
	workspaceLocks[workspaceRoot] = lock
	return lock
}

// SafeUpdateVideo updates a video record with automatic locking
func SafeUpdateVideo(ws *Workspace, videoID string, updates map[string]interface{}) (*ChannelManifest, error) {
	lock := GetWorkspaceLock(ws.Root)
	lock.mu.Lock()
	defer lock.mu.Unlock()

	// Load manifest
	manifest, err := ws.LoadManifest()
	if err != nil {
		return nil, err
	}

	// Find and update video
	var found bool
	for i := range manifest.Videos {
		if manifest.Videos[i].VideoID == videoID {
			// Apply updates
			if status, ok := updates["status"].(VideoStatus); ok {
				manifest.Videos[i].Status = status
			}
			if errorMsg, ok := updates["error"].(string); ok {
				manifest.Videos[i].Error = errorMsg
			}
			found = true
			break
		}
	}

	if !found {
		return nil, fmt.Errorf("video %s not found in manifest", videoID)
	}

	// Save manifest atomically
	if err := ws.SaveManifest(manifest); err != nil {
		return nil, err
	}

	return manifest, nil
}

// SafeLoadManifest loads manifest with read lock
func SafeLoadManifest(ws *Workspace) (*ChannelManifest, error) {
	lock := GetWorkspaceLock(ws.Root)
	lock.mu.Lock()
	defer lock.mu.Unlock()

	return ws.LoadManifest()
}

// SafeSaveManifest saves manifest with write lock
func SafeSaveManifest(ws *Workspace, manifest *ChannelManifest) error {
	lock := GetWorkspaceLock(ws.Root)
	lock.mu.Lock()
	defer lock.mu.Unlock()

	return ws.SaveManifest(manifest)
}
